src/aiida/tools/pytest_fixtures/daemon.py

```diff
@@ -116,7 +116,7 @@ def test(submit_and_await):
     from aiida.engine import ProcessState
 
     def factory(
-        submittable: 'Process' | 'ProcessBuilder' | 'ProcessNode',
+        submittable: type[Process] | 'ProcessBuilder' | 'ProcessNode' | t.Any,
```

Comment:
Why is `t.Any` needed here? This makes the type checking weaker, would be nice if we could avoid it.

Reply:
Because there is an else branch otherwise the branch is unreachable.

Reply:
Yes, makes sense, sorry I originally missed your comment with the errors.

I think the correct fix here is to add type ignores on those branches, since those essentially represent runtime type checking, which we need to enforce. At the same time we don't want to make static type checking weaker.

Reply:
I may not fully agree. I think types in the function signature plays two functionalities. On the one hand, it partially is the hint for the function behavior, like in this case in the function body it requires to explicitly deal with three types and raise if it doesn't know what to do with certain type. On the other hand, the type in the function signature is for other function to know what should be passed to the function when it is called. 
Ideally the `# type: ignore` should all be addressed. If we just rely on runtime type check, then the type should be `t.Any`.

Reply:
>  Ideally the # type: ignore should all be addressed.

I agree with this sentiment, but introducing `Any` is just another way to silence type checking. :-) I would argue that in this case `Any` is worse, since it "turns off" type checking of this parameter for the whole function, whereas the "type: ignore" would only target a specific part of the code. In this case, there is a real tension between static and runtime type checking: the else branches that are "unreachable" to the static type checker, but static types are not enforce at runtime in Python, and we can't enforce users to use mypy (e.g. in plugins that use this fixture), hence the additional runtime check.

> If we just rely on runtime type check, then the type should be t.Any

But we don't want to rely on runtime checks only. If we can catch it statically that's certainly better, no?

Sorry if I am being terse, I am teaching now, am happy to write more later today if needed.

Reply:
Right, I see your point now. I think you were suggest change to 

```
submittable: type[Process] | 'ProcessBuilder' | 'ProcessNode' # type: ignore
```

I thought you were suggest to not using type and just ignore it, sorry. I'll change this.

Reply:
Seems it is just works even without the `t.Any`. Please have a look.

Reply:
Oh no, my suggestion was to use the type ignores where mypy needs them, which is on the else branches I believe.

Reply:
Hmm, I don't think that's correct. :sweat_smile: 

I think the error goes away simply because you introduced a type that would fall through to the else branch. You can try that if you do this `Process | 'ProcessBuilder' | 'ProcessNode' | None` it would silence the error as well but would not be correct.

Reply:
btw: Thanks for doing this, fixing type issues is a very annoying work (at least to me :-) )

Reply:
So you suggest to what? 

Should be
```
-> type[Process] | 'ProcessBuilder' | 'ProcessNode' # type: ignore
```
?

Reply:
> Thanks for doing this, fixing type issues is a very annoying work 

True, for this part, I'd much prefer the strict type language. Doing this in python actually much more tedious.

Reply:
But with change it to 
```
type[Process] | 'ProcessBuilder' | 'ProcessNode'
```

The else branch didn't ask for type ignore by mypy anymore. So I think with `t.Any` is correct and better.

Reply:
I think the correct thing to do here is to NOT change the type annotation, since it was correct.

add type: ignore to the else branch, where the error originates.
The error is not really about type annotation, but about unreachable code. But we already know that branch is unreachable, unless the user screws up and passes a wrong type.

Reply:
`Process` -> `type[Process]` should be a correct change, in this function, it expect both workchain and calcjob which are subclass of `Process`. If it is just Process, then the `issubclass(submittable, Process)` part is useless because `and` shortcut the condition. 
We can have a zoom chat if you like, can be easier.

Reply:
In mega office, we decide to merge this one and continue the discussion here to not block other PRs. @danielhollas hope you don't mind.

Reply:
Sure, go ahead